app/controllers/bankdetails/BankAccountNameController.scala

```diff
@@ -39,46 +39,57 @@ class BankAccountNameController @Inject()(
+  def get(index: Option[Int] = None, edit: Boolean = false) = Authorised.async {
     implicit authContext =>
       implicit request =>
-        getData[BankDetails](index) map {
-          case Some(BankDetails(_, Some(data), _, _, _, _, _)) =>
-            Ok(views.html.bankdetails.bank_account_name(Form2[String](data), edit, index))
-          case Some(_) =>
-            Ok(views.html.bankdetails.bank_account_name(EmptyForm, edit, index))
-          case _ => NotFound(notFoundView)
-      }
+        index match {
+          case Some(i) =>
+            getData[BankDetails](i) map {
+              case Some(BankDetails(_, Some(data), _, _, _, _, _)) =>
+                Ok(views.html.bankdetails.bank_account_name(Form2[String](data), edit, Some(i)))
+              case Some(_) =>
+                Ok(views.html.bankdetails.bank_account_name(EmptyForm, edit, Some(i)))
+            }
+          case _ => Future.successful(Ok(views.html.bankdetails.bank_account_name(EmptyForm, edit, None)))
+        }
   }
 
-  def post(index: Int, edit: Boolean = false) = Authorised.async {
+  def post(index: Option[Int] = None, edit: Boolean = false) = Authorised.async {
     implicit authContext =>
       implicit request => {
         Form2[String](request.body) match {
           case f: InvalidForm =>
             Future.successful(BadRequest(views.html.bankdetails.bank_account_name(f, edit, index)))
           case ValidForm(_, data) =>
-            updateDataStrict[BankDetails](index) { bd =>
-              bd.copy(
-                accountName = Some(data),
-                status = Some(if (edit) {
-                  StatusConstants.Updated
+            val newBankDetails = BankDetails(accountName = Some(data))
+            index match {
+              case Some(i) => updateDataStrict[BankDetails](i) { bd =>
+                bd.copy(
+                  accountName = Some(data),
+                  status = Some(if (edit) {
+                    StatusConstants.Updated
+                  } else {
+                    StatusConstants.Added
+                  })
+                )
+              } map { _ =>
+                if (edit) {
+                  Redirect(routes.SummaryController.get(i))
                 } else {
-                  StatusConstants.Added
-                })
-              )
-            } map { _ =>
-              if (edit) {
-                Redirect(routes.SummaryController.get(index))
-              } else {
-                Redirect(routes.BankAccountIsUKController.get(index))
+                  Redirect(routes.BankAccountTypeController.get(i))
+                }
+              }
+              case _ => dataCacheConnector.fetch[Seq[BankDetails]](BankDetails.key) flatMap { maybeBankDetails =>
+                val newList = maybeBankDetails.getOrElse(Seq.empty) ++ Seq(newBankDetails)
+                dataCacheConnector.save(BankDetails.key, newList) map { _ => Redirect(routes.BankAccountTypeController.get(index.get)) }
```

Comment:
An issue I've just come across is that the index value that is passed to the `BankAccountTypeController.get` must be the index of the newly created instance of `BankDetails`. In this case, the index you want is the same value as `newList.size`, as it's just the index of the last inserted index.